File: c++/leetcode/958. Check Completeness of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
  bool isCompleteTree(TreeNode *root)
  {
    queue<TreeNode *> q;
    q.emplace(root);

    bool found_null = false;

    while (!q.empty())
    {
      const int sz = q.size();

      for (int i = 0; i < sz; ++i)
      {
        const auto front = q.front();

        if (found_null && front)
          return false;

        if (front)
        {
          q.emplace(front->left);
          q.emplace(front->right);
        }

        else
          found_null = true;

        q.pop();
      }
    }

    return true;
  }
};

File: javascript/leetcode/958. Check Completeness of a Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isCompleteTree = function (root) {
  let q = [root];
  let found_null = false;

  while (q.length) {
    const sz = q.length;

    for (let i = 0; i < sz; ++i) {
      const front = q[0];

      if (!front) found_null = true;
      else {
        if (found_null) return false;
        q.push(front.left);
        q.push(front.right);
      }

      q = q.slice(1);
    }
  }

  return true;
};
